raiden/transfer/mediated_transfer/initiator_manager.py

```diff
@@ -7,54 +7,58 @@
 from raiden.transfer.mediated_transfer.events import EventUnlockClaimFailed, EventUnlockFailed
 from raiden.transfer.mediated_transfer.state import (
     InitiatorPaymentState,
+    InitiatorTransferState,
     TransferDescriptionWithSecretState,
 )
 from raiden.transfer.mediated_transfer.state_change import (
-    ActionCancelRoute,
     ActionInitInitiator,
     ReceiveLockExpired,
     ReceiveSecretRequest,
     ReceiveSecretReveal,
     ReceiveTransferRefundCancelRoute,
 )
-from raiden.transfer.state import NettingChannelState
+from raiden.transfer.state import RouteState
 from raiden.transfer.state_change import ActionCancelPayment, Block, ContractReceiveSecretReveal
-from raiden.utils.typing import BlockNumber, ChannelMap, List
+from raiden.utils.typing import BlockNumber, ChannelMap, List, SecretHash, cast
 
-# TODO:
-# - Add synchronization for expired locks (issue #193).
-#   Transfers added to the canceled list by an ActionCancelRoute are stale in
-#   the channels merkle tree, while this doesn't increase the messages sizes
-#   nor does it interfere with the guarantees of finality it increases memory
-#   usage for each end, since the full merkle tree must be saved to compute
-#   it's root.
 
+def clear_if_finalized(iteration: TransitionResult) -> TransitionResult:
+    """ Clear the initiator payment task if all transfers have been finalized
+    or expired. """
+    state = cast(InitiatorPaymentState, iteration.new_state)
 
-def iteration_from_sub(
-        payment_state: InitiatorPaymentState,
-        iteration: TransitionResult,
-) -> TransitionResult:
+    if state is None:
+        return iteration
+
+    if len(state.initiator_transfers) == 0:
```

Comment:
this is not taking care of the cancelled routes, which have to be cleared too, otherwise we will have expired locks in the merkle tree

Reply:
There's no `cancelled_transfers/routes` vs `pending_transfers`. We only have 1 list `initiator_transfers` which holds all transfers. Each transfer as a `transfer_state` which can be either `pending` or `cancelled`. So this would check if ALL transfers pending/completed/cancelled are removed then the parent initiator manager task can be cleared.

Reply:
Same thing, initiators are not cleared from the map when cancelled.